proofs/tree.go

```diff
@@ -327,14 +326,9 @@ func (doctree *DocumentTree) CreateProof(prop string) (proof proofspb.Proof, err
 	if leaf == nil {
 		return proofspb.Proof{}, fmt.Errorf("No such field: %s in obj", prop)
 	}
-	var propBytes []byte
-	if doctree.compactProperties {
-		propBytes = leaf.Property.CompactName()
-	} else {
-		propBytes = []byte(leaf.Property.Name())
-	}
+	propName := leaf.Property.Name(doctree.compactProperties)
 	proof = proofspb.Proof{
-		Property: propBytes,
+		Property: propName.AsBytes(),
```

Comment:
Or it could be a LiteralPropName which is just a []byte. This would happen if we were calculating a hash from a Proof object. The Proof object has a []byte, not a PropertyName, so LiteralPropName is a simple way to create a PropertyName from the []byte (which gets converted back to a []byte on this line). We could store a PropertyName in the Proof object instead if you wanted to get rid of LiteralPropName.

Reply:
Not sure I'm following your explanation completely. The `Proof` struct is the export format (serialized to JSON to transmit in an API response for example).

It is used to by a third party concatenate the values of the leaf `(property, value, salt)` and validate the Merkle proof of the resulting hash against a known Merkle root.

 If the tree was constructed with the compact property names, the `Proof` needs to contain the compact property attribute. If it's using human readable properties, it should contain the string. As this value is needed when concatenating the leaf preimage.

I think the most elegant way would be to extend the `Property` protobuf message to have both `StringProperty` and `CompactProperty` or something similar and fill either value depending on the tree type.

Reply:
> If the tree was constructed with the compact property names, the Proof needs to contain the compact property attribute. If it's using human readable properties, it should contain the string. As this value is needed when concatenating the leaf preimage.

Right now, the Proof contains only the bytes. In order to validate the Proof, one needs to know whether it is a compact or human-readable properties Proof. Everything currently works using that.

The question now is one that I can't answer because I don't have enough domain experience here, so I am fine doing whichever you prefer: should the Proof indicate whether its using compact properties or human-readable properties? Or should this be knowledge that is already known?

Reply:
added the oneof property field to the proof protobuf